File: Palindromo/Palindromo.java
import java.util.Scanner;

class Palindromo {

   public static boolean isFim(String s) {
      return (s.length() >= 3 && s.charAt(0) == 'F' && s.charAt(1) == 'I' && s.charAt(2) == 'M');
   }

   public static boolean ePalindromo(String palavra) {
      String contrario = "";
      for (int i = (palavra.length() - 1); i >= 0; i--) {
         contrario = contrario + palavra.charAt(i);
      }
      
      int dif = 0;
      for(int i = 0; i < palavra.length(); i++){
         if(palavra.charAt(i) > contrario.charAt(i) || palavra.charAt(i) < contrario.charAt(i)) 
            dif = 1;
      } 
      if(dif == 1)
         return false;
      else return true;
         
   }

   public static void main(String[] args) {
      String[] entrada = new String[1000];
      int numEntrada = 0;
      Scanner entrada2 = new Scanner(System.in);

      // Leitura da entrada padrao
      do {
         entrada[numEntrada] = entrada2.nextLine();
      } while (isFim(entrada[numEntrada++]) == false);
      numEntrada--; // Desconsiderar ultima linha contendo a palavra FIM
      entrada2.close();

      for (int i = 0; i < numEntrada; i++) {
         if (ePalindromo(entrada[i]) == true)
            System.out.println("SIM");
         else
            System.out.println("NAO");
      }
   }
}

File: Palindromo/palindromo.cpp
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <iostream>
#include <locale.h>
using namespace std;
#define bool short
#define true 1
#define false 0

bool isFim(char *s){
    return (strlen(s) >= 3 && s[0] == 'F' &&
        s[1] == 'I' && s[2] == 'M');
}

char *replaceChar(char *string, char toSearch, char toReplace)
{
    char *charPtr = strchr(string, toSearch);
    if (charPtr != NULL) *charPtr = toReplace;
    return charPtr;
}

void lerLinha(char *string, int tamanhoMaximo, FILE *arquivo)
{
    fgets(string, tamanhoMaximo, arquivo);
    replaceChar(string, '\r', '\0');
    replaceChar(string, '\n', '\0');
}

bool palindromo(char *s) {
    char comp[1000];
    int tam = strlen(s) - 1;
    bool resp = 1;
    int i, j;
    for(i = 0, j = strlen(s)-1; i < strlen(s); i++, j--){
        comp[i] = s[j];
    }

    for(i = 0; i < strlen(s)-1; i++){
        if(s[i] == comp[i]){
            resp = 1;
        }
        else{
            resp = 0;
            i = strlen(s);
        }  
    }
    return resp;
}

int main() {
    setlocale (LC_ALL, "");
    char entrada [700][1000];
    int numEntrada = 0;

    // Leitura da entrada padrao
    do{
        lerLinha(entrada[numEntrada], 1000, stdin);
    }while (isFim(entrada[numEntrada++]) == false);
    numEntrada--;

    int i;
    for(i = 0; i < numEntrada; i++){
        int resp = palindromo(entrada[i]);
        if(resp == true)
            cout << "SIM" << endl;
        else 
            cout << "NAO" << endl;
    }
}


